File: php/MongoMoviesAggregationTest.php
<?php
require_once ('PHPUnit/Framework/Assert.php');
require_once ('PHPUnit/Framework/TestCase.php');


class MongoMoviesAggregationTest extends PHPUnit_Framework_TestCase {

    private $mongoclient;
    private $db;
    private $movies;

    public function setUp(){
      $this->mongoclient = new MongoClient();
      $this->db = $this->mongoclient->selectDB("mediatheque");
      $this->movies = $this->db->selectCollection("movies");
    }

    public function tearDown(){
        $this->mongoclient->close();
    }

    /**
     * Trouver l'actor avec le plus grand nombre de films à son actif
     */
    public function testFindActorWithLargestPorfolio(){
        // TODO définir le pipeline

        $result = $this->zips->aggregate();

        $this->assertEquals('Robert De Niro', $result['result'][0]['_id']);
        $this->assertEquals(9, $result['result'][0]['number_of_movies']);
    }

    /**
     * Trouver l'actor avec le plus grand nombre de films à son actif
     * avec un map reduce
     */
    public function testFindActorWithLargestPorfolioWithMapReduce(){
        $map = null;
        $reduce = null;

        $this->db->command(array(
            "mapreduce" => "movies",
            "map" => $map,
            "reduce" => $reduce,
            "out" => "movies_per_actor"));

        $movies_per_actor = $this->db->selectCollection("movies_per_actor");
        $query = array();
        $this->assertTrue($movies_per_actor->count($query)>0);
        $actors = $movies_per_actor->find()->sort(array('value.count'=>-1))->limit(1);
        foreach($actors as $actor) {
            $this->assertEquals('Robert De Niro', $actor['_id']);
            $this->assertEquals(9, $actor['value']['count']);
        }
    }
}
?>

File: python/MongoMoviesAggregationTest.py
# -*- coding: utf-8 -*-
import unittest
import pymongo
from pymongo import MongoClient
from bson.code import Code


class MongoMoviesAggregationTest(unittest.TestCase):

    #
    # Initialisation des membres privés à compléter
    def setUp(self):
        self.mongoclient = MongoClient()
        self.db = self.mongoclient.mediatheque
        self.movies = self.db.movies


    def tearDown(self): 
        self.mongoclient.close()

    #  Trouver l'actor avec le plus grand nombre de films à son actif
    def testFindActorWithLargestPorfolio(self):
        unwind = {'$unwind' : '$actors' }
        group = {"$group" : {
                                '_id' : '$actors', 'number_of_movies' : {'$sum' : 1}
                            }
        }
        sort = {"$sort" : {"number_of_movies" : -1}}
        limit = {"$limit" : 1}
        result = self.movies.aggregate([unwind, group, sort, limit])
        self.assertEqual(result['result'][0]['_id'], 'Robert De Niro')
        self.assertEqual(result['result'][0]['number_of_movies'], 9)

    #  Trouver l'actor avec le plus grand nombre de films à son actif
    def testFindActorWithLargestPorfolioWithMapReduce(self):
        mapper = Code("""
                       function() {
                            for (var idx = 0; idx < this.actors.length; idx++) {
                                var key = this.actors[idx];
                                emit(key, this.title);
                            }
                        }
                       """)
        reducer = Code("""
                       function(actor, movies) {
                            movie_list = {count : 0, films : []};
                            for (var i in movies) {
                                movie_list.films = movies[i].concat(movie_list.films);
                                movie_list.count++;
                            }

                            return movie_list;
                        }
                       """)

        result = self.db.movies.map_reduce(mapper, reducer, "movies_per_actor")
        self.assertEqual(result.find().count(), 4346)
        actors = self.db.movies_per_actor.find().sort([("value.count", -1)]).limit(1)
        for actor in actors:
            self.assertEqual('Robert De Niro', actor['_id'])
            self.assertEqual(9, actor['value']['count'])


if __name__ == '__main__':
    unittest.main()
